src/v/cluster/controller.cc

```diff
@@ -548,59 +574,45 @@ ss::future<> controller::create_cluster() {
-    if (!is_cluster_founder) {
-        co_return;
-    }
+    const bool controller_log_exists = _raft0->last_visible_index()
+                                       > model::offset{};
 
-    if (config::node().empty_seed_starts_cluster()) {
-        // This is the legacy empty-seed-starts-cluster mode
+    if (co_await discovery.is_cluster_founder()) {
         vassert(
-          config::node().seed_servers().empty(), "We are not on the root node");
-
-        if (
-          _raft0->last_visible_index() > model::offset{}
-          || _raft0->config().brokers().size() > 1) {
-            // The controller log has already been written to
-            co_return;
+          !controller_log_exists,
+          "Can't be a cluster founder with controller log present");
+        if (config::node().empty_seed_starts_cluster()) {
+            vassert(
+              config::node().seed_servers().empty(),
+              "Cluster founder is not the root node");
+            vassert(
+              _raft0->config().brokers().size() == 1,
+              "Root is a cluster founder but joiners are in the cluster alrdy");
         }
 
-        // Internal RPC does not start until after controller startup
-        // is complete (we are called during controller startup), so
-        // it is guaranteed that if we were single node/empty controller
-        // log at start of this function, we will still be in that state
-        // here.  The wait for leadership is really just a wait for the
-        // consensus object to finish writing its last_voted_for from
-        // its self-vote.
-        while (!_raft0->is_leader()) {
-            co_await ss::sleep(100ms);
-        }
-
-        co_return co_await create_cluster();
-    }
-
-    // The new cluster creation is done by the leader of raft0 consisting of
-    // seed servers, once elected
-    while (!_storage.local().get_cluster_uuid() && !_raft0->is_leader()) {
-        co_await ss::sleep(100ms);
+        co_return co_await create_cluster(create_cluster_mode::bootstrap);
     }
 
-    if (_storage.local().get_cluster_uuid()) {
+    if (controller_log_exists && !_storage.local().get_cluster_uuid()) {
+        // No cluster UUID while controller log present - we are upgrading
+        // an existing cluster
         vlog(
           clusterlog.info,
-          "Cluster is {}",
-          *_storage.local().get_cluster_uuid());
-        co_return;
+          "Cluster will be upgraded once all nodes are upgraded. "
+          "Root driven bootstrap: {}",
+          config::node().empty_seed_starts_cluster());
+        ssx::background
+          = _feature_table.local()
+              .await_feature(
+                features::feature::seeds_driven_bootstrap_capable, _as.local())
+              .then([this] {
+                  return create_cluster(create_cluster_mode::cluster_uuid_only);
+              })
+              .then([] { vlog(clusterlog.info, "Cluster has been upgraded"); })
```

Comment:
nit: similar to above, as someone just reading this log without context into this project, I think it won't be clear what this means. Let's focus this message on what we are actually doing, maybe something like:

"Finished writing cluster UUID to controller"

Reply:
logging has been refactored